File: go/1049M-最后一块石头的重量II.go
/*
 * @Author: SukiEva
 * @Date: 2021-11-29 10:33:25
 * @Description: https://github.com/SukiEva
 * わたし、気になります！
 */

package main

func lastStoneWeightII(stones []int) int {
	sum := 0
	for _, v := range stones {
		sum += v
	}
	tmp := sum
	sum /= 2
	dp := make([]int, sum+1)
	for i := range stones {
		for j := sum; j >= stones[i]; j-- {
			dp[j] = max(dp[j], dp[j-stones[i]]+stones[i])
		}
	}
	return tmp - dp[sum]-dp[sum]
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

File: cpp/1049M-最后一块石头的重量II.cpp
/*
 * @Author: SukiEva
 * @Date: 2022-01-10 12:35:37
 * @Description: https://github.com/SukiEva
 * わたし、気になります！
 */
#include <bits/stdc++.h>

#include "include/ListNode.h"
#include "include/TreeNode.h"

using namespace std;

class Solution {
   public:
    int lastStoneWeightII(vector<int>& stones) {
        int sum = 0;
        for (int stone : stones) sum += stone;
        int tmp = sum;
        sum >>= 1;
        vector<int> dp(sum + 1);
        for (int i = 0; i < stones.size(); i++) {
            for (int j = sum; j >= stones[i]; j--) {
                dp[j] = max(dp[j], dp[j - stones[i]] + stones[i]);
            }
        }
        return tmp - 2 * dp[sum];
    }
};